AE-1987: käskypäätös / varsinainen päätös toimenpide changes

Juholei opened the pull request:
- Move käskypäätös / varsinainen päätös toimenpide fields 'recipient-answered', 'answer-commentary-fi', 'answer-commentary-sv', 'statement-fi', 'statement-sv' to be osapuoli-specific

- New UI where all the osapuoli specific fields are organized by osapuoli

Juholei committed:
AE-1987: Move käskypäätös / varsinainen päätös toimenpide fields 'recipient-answered', 'answer-commentary-fi', 'answer-commentary-sv', 'statement-fi', 'statement-sv' to be osapuoli-specific

- Preliminary UI changes to support this change included, will be rethought later

Juholei committed:
AE-1987: Simplify osapuoli name formatting in new-actual-decision-toimenpide.svelte

Juholei committed:
AE-1987: If recipient didn't answer, related fields are marked as optional in schema and hidden in UI for käskypäätös / varsinainen päätös toimenpidetype

Juholei committed:
AE-1987: Set answer-commentary-fi, answer-commentary-sv, statement-fi, statement-sv to null if recipient-answered is false when sending toimen pide to backend

Juholei committed:
AE-1987: Clear käskypäätös / varsinainen päätös answer-commentary and statement fields when there is no document for osapuoli, move their original schema under osapuoli-specific-data

Juholei committed:
AE-1987: Remove keys with null values from osapuoli-specific-data when serializing it

Juholei committed:
AE-1987: Remove recipient-answered field in käskypäätös / varsinainen päätös if the osapuoli doesn't have document

Juholei committed:
AE-1987: Implement new ui for käskypäätös / varsinainen päätöstoimenpide

- All osapuoli specific fields are under the heading with the osapuoli name
- Separate osapuolitable removed as unnecessary

Juholei committed:
AE-1987: Add spinner when preview is pending in the new käskypäätös / varsinainen päätös toimenpide ui

Juholei committed:
AE-1987: Send only the osapuoli-specific-data of the current osapuoli in the preview api call for käskypäätös / varsinanen päätös document

- If other osapuolis are still missing the data, it's now possible to preview the document for an osapuoli that has all the required data filled in

Juholei committed:
AE-1987: Show validation errors in new-actual-decision-toimenpide subview for each osapuoli

- Previously the errors were visible only in the top of the dialog so they weren't visible most of the time

Juholei commented on src/pages/valvonta-kaytto/new-actual-decision-toimenpide.svelte:
```diff
@@ -5,86 +5,79 @@
-      validators={R.path(['type-specific-data', 'statement-sv'], schema)}
-      {i18n} />
-  </div>
+<style type="text/postcss">
```
Ehkä tän voisi refaktoroida joksikin ErrorContainer-komponentiksi ja sitten käyttää sitä sekä Dialog-komponentissa (mistä apinoitu tähän) että tässä?

Juholei committed:
AE-1987: Rename courtDataIndexForOsapuoli to osapuoliSpecificDataIndexForOsapuoli

- Matches the current functionality and usage better

Juholei commented on src/pages/valvonta-kaytto/valvonta-api.js:
```diff
@@ -232,22 +232,53 @@ const setFieldToNoneIfNoDocument = field => data => {
 const unwrapMaybeIfExists = field => data => {
   return R.when(
-    R.has(field),
+    R.both(R.has(field), R.compose(Maybe.isMaybe, R.view(R.lensProp(field)))),
```
Olisiko tässä riittänyt R.view(R.lensProp(field)) sijaan R.prop(...

solita-antti-mottonen replied:
Pitäisikö tästä muutoksesta tehdä eri funktio?

Juholei committed:
AE-1987: Remove outdated comments

solita-antti-mottonen commented:
Joku erotin loppuun ennen mainintaa manuaalisesta toimittamisesta olisi hyvä

Juholei commented on src/pages/valvonta-kaytto/new-actual-decision-toimenpide.svelte:
```diff
@@ -138,4 +131,206 @@
+          <div class="w-full py-4">
+            <Select2
+              bind:model={toimenpide}
+              label="Hallinto-oikeus"
```
Siirrä kielistyksiin

solita-antti-mottonen commented on src/pages/valvonta-kaytto/new-actual-decision-toimenpide.svelte:
```diff
@@ -5,86 +5,79 @@
+      osapuoli.id
+    );
+    preview(
+      osapuoli.type.preview(id, osapuoli.id, previewToimenpide),
```
Jos tämä parametri olisikin funktio joka ottaa tuon `previewToimenpiteen` myöhemmin niin ei jäisi mahdollisuutta validoida eri asiaa kuin mitä lähetetään

Juholei replied:
Tämä tuntuu äkkiseltään liian isolta remontilta tähän palaan

solita-antti-mottonen commented on src/pages/valvonta-kaytto/schema.js:
```diff
@@ -121,7 +109,46 @@ export const toimenpidePublish = (templates, toimenpide) =>
+              ],
+              toimenpide
+            );
+            // TODO: Siisti toisteisuus
```
Koska?

solita-antti-mottonen commented on src/pages/valvonta-kaytto/valvonta-api.js:
```diff
@@ -232,22 +232,53 @@ const setFieldToNoneIfNoDocument = field => data => {
+      setFieldToNoneIfNoDocument('answer-commentary-fi'),
+      setFieldToNoneIfNoDocument('answer-commentary-sv'),
+      setFieldToNoneIfNoDocument('statement-fi'),
+      setFieldToNoneIfNoDocument('statement-sv'),
```
Näitä voisi ehkä kasata composella yhden kentän funktioiksi ja loopata sitten kaikille kentille mille se operaatio tehdään?

Juholei replied:
Makustelin tätä ja minusta tuntui paremmalta tehdä niin päin, että tein näille funktioille wrapperit jotka ottavat parametrina listan fieldejä ja sitten ajavat niille kaikille nämä alkuperäiset funktiot. Tehtävät operaatiot pysyivät selkeinä ja sitten helpompi määritellä mille kentille tehdään mikäkin. Miltä näyttää sun mielestä tämä toteutus?

solita-antti-mottonen replied:
Selkeämpi ehdottomasti kuin alkuperäinen. Mennään tällä

solita-antti-mottonen commented on src/pages/valvonta-kaytto/schema.js:
```diff
@@ -121,7 +109,46 @@ export const toimenpidePublish = (templates, toimenpide) =>
+            );
+            // TODO: Siisti toisteisuus
             return R.compose(
+              R.over(
```
Nämä voisi varmaan loopata kun kaikille tehdään sama operaatio?

solita-antti-mottonen commented on src/pages/valvonta-kaytto/valvonta-api.js:
```diff
@@ -232,22 +232,53 @@ const setFieldToNoneIfNoDocument = field => data => {
     R.compose(
+      removeNullValues,
+      unwrapMaybeIfExists('recipient-answered'),
+      setFieldToNoneIfNoDocument('recipient-answered'),
```
Tähän tilalle `setFieldToNullIfNoDocument`

Juholei committed:
AE-1987: Fix hallinto-oikeus selection label to come from kielistykset

Juholei committed:
AE-1987: Clean repetitive code in serializeOsapuoliSpecificData

Juholei committed:
AE-1987: Refactor validation errors in Dialog and new-actual-decision-toimenpide view to a new Error component

Juholei committed:
AE-1987: Create a new schema based on the toimenpide data being validated for preview

- Otherwise the partial data was validated agains a full schema that expected all osapuolis to be present

Juholei committed:
AE-1987: Simplify käskypäätös / varsinainen päätös schema creation

Juholei committed:
AE-1987: Add border to separate the info text for manual sending from the fields in käskypäätös / varsinainen päätös form
